### PR TITLE
add feature to report bad comments

### PR DESCRIPTION
**Pull request by nif33**
add a report button on every comment
Upon clicking button, a fetch is made to the database to increase the number of reports by one
if a comment is reported twice it will not be shown to users

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -50,13 +52,17 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
     datastore.put(commentEntity);
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    Query query = new Query("Comment");
+    query.addSort("numReports", SortDirection.ASCENDING); // required sorting to apply filter
+    query.addSort("timestamp", SortDirection.DESCENDING); // display in order of recency
+    query = query.addFilter("numReports", Query.FilterOperator.LESS_THAN, 2); // only show comments with less than 2 reports
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Comments are redundant because the code is self explanatory. Check out the `To Comment or Not to Comment?` ToTT.

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+public class ReportServlet extends HttpServlet {
+
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String id = request.getParameter("id");
+    try {
+      Key commentKey = KeyFactory.stringToKey(id);
+  		Entity commentEntity = datastore.get(commentKey);
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: Bad indentation.

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      long numReports = (long) commentEntity.getProperty("numReports");
+      numReports++;
+      commentEntity.setProperty("numReports", numReports);
+      datastore.put(commentEntity);
+  	} catch (EntityNotFoundException e) {
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: Minimize code inside try block. The first two lines should be enough.

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String id = request.getParameter("id");
+    try {
+      Key commentKey = KeyFactory.stringToKey(id);
+  		Entity commentEntity = datastore.get(commentKey);
+      long numReports = (long) commentEntity.getProperty("numReports");
+      numReports++;
+      commentEntity.setProperty("numReports", numReports);
+      datastore.put(commentEntity);
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: Andrew in a different PR. Replace `like` with `report`
> What happens if two people have your website open and like the same comment at the exact same time? ;)

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -35,18 +35,31 @@ function newCommentBox(comment){
...
   const commentText = document.createElement('p');
+  const commentReport = document.createElement('a');
 
-  // set text for elements
+  // set info for elements
   commentName.innerText = comment.name + ':';
   commentText.innerText = comment.text;
+  commentReport.innerText = "Report";
+  commentReport.onclick = function() {
+    addReport(comment.id);
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js:45]: Nit: In the back-end you are adding reports but in the front-end to the users they are reporting a comment. I recommend renaming to reportComment or something along those lines.

**Review by juanalonsocruz — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java (hunk 2 above)
Comment (nif33) [reply portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: weird. Looked normal in Atom, must be some setting difference in tab spacing. I'll make sure to check the Github display next time.

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java (hunk 4 above)
Comment (nif33) [reply portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: I actually talked about this yesterday with Zoe. I feel like it's less important for my use case because I am counting report numbers and I arbitrarily chose 2 reports to be the number required for a comment to be taken down (it'd have to be more if the website had more traffic).
Nonetheless, I have amended my code according to this documentation: https://cloud.google.com/appengine/docs/standard/java/datastore/transactions#using_transactions
1) I accidentally amended onto the wrong commit (oops, forgot I did a merge, will make sure to check next time)
2) the documentation code uses an infinite loop and breaks from it, in class I've always learned that using infinite loops is bad practice, what is the truth? I feel like it might be less readable if I did numRetries in a for loop...

Comment (juanalonsocruz) [reply portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: In general, you want to have accurate data for future use cases. What if in the future you want to create a new page that displayed metrics that depend directly on the number of reports?

For your (2) comment, I agree. for loop is good. I would also put everything that inside the loop into a method called something like `incrementReportCount(...)`

Overall: Please don't spend much more time on this. It is mainly a good thing to think about for capstone.

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        try {
+          String id = request.getParameter("id");
+          Key commentKey;
+          Entity commentEntity;
+          try {
+            commentKey = KeyFactory.stringToKey(id);
+            commentEntity = datastore.get(commentKey);
+        	} catch (EntityNotFoundException e) {
+        		throw new RuntimeException("Key is not an entity in the database");
+        	}
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: Indentation. Wonder whats happening with Atom and github

Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: No need to catch and throw a similar exception. `e` possibly has more information in its message about what went wrong.

**Review by juanalonsocruz — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 7]
```diff
@@ -22,37 +22,39 @@
...
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      try {
+        Key commentKey = KeyFactory.stringToKey(id);
+        Entity commentEntity = datastore.get(commentKey);
+        long numReports = (long) commentEntity.getProperty("numReports");
+        numReports++;
+        commentEntity.setProperty("numReports", numReports);
+        datastore.put(transaction, commentEntity);
+        transaction.commit();
```
Comment (nif33) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: I moved this back into the try/catch because it was giving me errors related to the commentEntity might not be initiated. I will go back to this issue at the end of the week if I have time so I don't spend more time on this part of the project right now.

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 8]
```diff
@@ -22,37 +22,39 @@
...
-            throw e;
-          }
-          numRetries--;
-        } finally {
-          if(transaction.isActive()) {
-            transaction.rollback();
-          }
-        }
+    while(numRetries > 0) {
+      numRetries = incrementReportCount(id, numRetries);
```
Comment (nif33) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: I used a while loop instead of a for because the function has to return numRetries, and it seemed redundant to update the numRetries in a for loop

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 9]
```diff
@@ -22,37 +22,39 @@
 
   DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
+  public int incrementReportCount(String id, int numRetries) {
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      try {
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: No need to have cascading try blocks. Java lets you catch different exceptions like this

```
try {
  ...
} catch (Exception1 e) {
  ...
} catch (Exception2 e) {
  ...
} finally {
  ...
}
```

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java (hunk 8 above)
Comment (juanalonsocruz) [reply portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: You could instead move the try catches outside the method into where the loop is. This way the helper method will be independent of retries.

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 10]
```diff
@@ -22,37 +22,39 @@
...
+        Key commentKey = KeyFactory.stringToKey(id);
+        Entity commentEntity = datastore.get(commentKey);
+        long numReports = (long) commentEntity.getProperty("numReports");
+        numReports++;
+        commentEntity.setProperty("numReports", numReports);
+        datastore.put(transaction, commentEntity);
+        transaction.commit();
+      } catch (EntityNotFoundException e) {
+        System.err.println("Key is not an entity in the database");
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: You should also inform the API client that something went wrong. Look into HTTPServletResponse.sendError(.,.)

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 11]
```diff
@@ -22,37 +22,39 @@
...
+      }
+      return 0;
+    } catch (ConcurrentModificationException e) {
+      if(numRetries == 0) {
+        throw e;
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: Instead of throwing e back. You should call HTTPServletResponse.sendError(.,.).

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -127,32 +127,6 @@ function addComment() {
...
-   }
-
-   const commentContainer = document.getElementById('comment-container');
-   const request = new Request('/delete-data', {method: 'POST'});
-
-   fetch(request).then(() => {
-     clearChildren(commentContainer);
-   });
- }
-
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js:142]: Do you intend to remove the delete all functionality? If so, this should be in its own PR.

**Review by juanalonsocruz — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 12 above)
Comment (nif33) [reply portfolio/src/main/webapp/script.js:142]: No, it was just some duplicate code that appeared, I think it added automatically during some merge request.
The file is getting kind of long and hard to manage, am I allowed to add new js files to separate things?

Comment (juanalonsocruz) [reply portfolio/src/main/webapp/script.js:142]: Yes! Anything to make it easier to read.

File: portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java [hunk 13]
```diff
@@ -22,39 +22,39 @@
...
-      }
-      return --numRetries;
-    } finally {
-      if(transaction.isActive()) {
-        transaction.rollback();
-      }
+  public boolean incrementReportCount(String id, Transaction transaction) throws EntityNotFoundException, ConcurrentModificationException {
+    Key commentKey = KeyFactory.stringToKey(id);
+    Entity commentEntity = datastore.get(commentKey);
+    if(commentEntity == null) {
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java]: get doesn't return null according to the doc. It would throw an exception. No need to do this check or have this method return a boolean since its not being used in the caller.
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#get-com.google.appengine.api.datastore.Key-

Also, the transaction is only grouping the put. You want to group both the get and write. Use get(transaction, commentKey) instead.
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#get-com.google.appengine.api.datastore.Transaction-com.google.appengine.api.datastore.Key-

**Review by juanalonsocruz — APPROVED** [review]
Small changes needed but otherwise approve =D

**Review by ajpaon — APPROVED** [review]
